src/impy/models/sophia.py

```diff
@@ -41,118 +44,65 @@ class Sophia20(MCRun):
 
     _name = "Sophia"
     _version = "2.0"
-    _event_class = SophiaEvent
     _library_name = "_sophia"
-    _output_frame = "laboratory"
-
-    def __init__(self, event_kinematics, seed="random", logfname=None):
-        super().__init__(seed, logfname)
+    _event_class = SophiaEvent
+    _frame = EventFrame.FIXED_TARGET
+    _projectiles = {lp.photon.pdgid}
+    _targets = {lp.p.pdgid, lp.n.pdgid}
+    _ecm_min = 0
 
-        self._lib.s_plist.ideb = impy_config["sophia"]["debug_level"]
+    def __init__(self, kinematics, *, seed=None, keep_decayed_particles=True):
+        import impy
 
-        self._lib.init_rmmard(self._seed)
+        super().__init__(seed)
 
+        self._lib.s_plist.ideb = impy.debug_level
         # Keep decayed particles in the history:
-        self._lib.eg_io.keepdc = impy_config["sophia"]["keep_decayed_particles"]
-        self.event_kinematics = event_kinematics
+        self._lib.eg_io.keepdc = keep_decayed_particles
+
+        self.kinematics = kinematics
         self._set_final_state_particles()
 
-    @staticmethod
-    def _validate_kinematics(k):
-        if k.p1pdg != 22:
-            raise ValueError(
-                "Sophia accepts only 'gamma' as a projectile "
-                + "but received pdg_id = {0}".format(k.p1pdg)
-            )
-
-        if k.p2pdg not in [2212, 2112]:
-            raise ValueError(
-                "Sophia accepts only 'proton' or 'neutron' as a target, "
-                + "but received pdg_id = {0}".format(k.p2pdg)
-            )
-
-        # HD: is this necessary?
-        if k.p2_is_nucleus:
-            raise ValueError(
-                "Sophia accepts only 'proton' or 'neutron' as a target, "
-                + "but received nucleus = ({0}, {1})".format(k.A2, k.Z2)
-            )
-
-    def _cross_section(self, evt_kin):
-        self._validate_kinematics(evt_kin)
-
-        nucleon_code_number = self._lib.icon_pdg_sib(evt_kin.p2pdg)
-        energy_of_photon = self.event_kinematics.elab
-
-        # self.energy_of_photon is the energy in lab frame
-        # where nucleon is at rest and photon is moving
-        total_crossection_id = 3  # 3 is for total crossection
-        # cross section in micro barn
-        total = self._lib.crossection(
-            energy_of_photon, total_crossection_id, nucleon_code_number
+    def _cross_section(self):
+        # code=3 for inelastic cross-section
+        # TODO fill more cross-sections
```

Comment:
The cross sections for photo-hadronic processes are different from the usual ones defined in the CrossSectionData (resonances, multipion, direct, etc.). Do you want to keep the class generic then there is no need to do more. Any preferences?

Reply:
I would keep it as it is.

impy is primarily a frontend for h-h, h-A, A-A interactions. This means the CrossSectionData should cover those cases. It is useful to further split the inelastic cross-section into non-diffractive and diffractive parts. g-h and g-A are also interesting to support, but I don't think we need to split the inelastic cross-section into parts for these interactions.